### PR TITLE
ARROW-5579: [JAVA][WIP] shade flatbuffer dependency version 2

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-5579](https://issues.apache.org/jira/browse/ARROW-5579), second try.

### CODE REVIEW

File: java/format/pom.xml [hunk 1]
```diff
@@ -159,6 +159,32 @@
...
+      <executions>
+        <execution>
+          <phase>package</phase>
+          <goals>
+            <goal>shade</goal>
+          </goals>
+          <configuration>
+            <artifactSet>
+              <includes>
+                <include>com.google.flatbuffers:*</include>
```
Comment (praveenbingo) [inline java/format/pom.xml]: how about excluding the flatbuffer builder? it seems like the only class that we use in Arrow.

**Comment by emkornfield** [comment]
@praveenbingo nice observation! Would it make sense to have a wrapper for it in format so it doesn't need to be excluded from the shading?

**Comment by tianchen92** [comment]
Cool, I think having a wrapper seems reasonable since we only use one class, I will have a try later, thanks!

**Comment by emkornfield** [comment]
I guess the one issue with a wrapper might be if it breakscapi compatibility

**Comment by tianchen92** [comment]
@emkornfield You are right, it has compatibility problems with a wrapper.

**Comment by emkornfield** [comment]
@praveenbingo @tianchen92 if we can't wrap the class I think we probably need to shade it  (or wrap more underlying flatbuffer infrastructure).  Is this accurate?

**Comment by tianchen92** [comment]
@emkornfield I have made a poc to create FlatbufferBuilder wrapper, see https://github.com/tianchen92/arrow/commits/ARROW-5579-testWrapper. The build works fine on my local machine and more important, it has no problem with IDE which the shaded approach has. Please take a look and see if it's reasonable, thanks!

**Comment by emkornfield** [comment]
> I have made a poc to create FlatbufferBuilder wrapper, see https://github.com/tianchen92/arrow/commits/ARROW-5579-testWrapper. The build works fine on my local machine and more important, it has no problem with IDE which the shaded approach has. Please take a look and see if it's reasonable, thanks!

I took a quick look, it seems that we added a dependency on flatbuffers in vector/pom.xml, I think this would leak flatbuffers to consumers?

**Comment by tianchen92** [comment]
@emkornfield Without this dependency, build will fail with:
[WARNING] Used undeclared dependencies found:
[WARNING]    com.google.flatbuffers:flatbuffers-java:jar:1.9.0:compile
What about relocate flatbuffers in vector which are not expected to be used by user anymore (we expect they use FlatBufferBuilderWrapper instead)?

**Comment by emkornfield** [comment]
> If all arrow interfaces exposed to users are replaced with wrapper class, it won't cause version conflict?

@tianchen92 I don't think this is true.  Only one version of the flatbuffers class can be loaded at a time, and that would cause incompabilities

**Comment by tianchen92** [comment]
Thank you @emkornfield, do you have any suggestion? So far we have tried two approaches(relocate class and wrapper), but none of them solved the problem perfectly.

**Comment by emkornfield** [comment]
@tianchen92 we could potentially move stuff under types.pojo to format and create a wrapper around FBSerializable and ipc.message to format and that might work, I havne't looked closely at it though.

**Comment by emkornfield** [comment]
probably need to do something about FBSerializable interface as well.

**Comment by tianchen92** [comment]
> > If all arrow interfaces exposed to users are replaced with wrapper class, it won't cause version conflict?
> 
> @tianchen92 I don't think this is true. Only one version of the flatbuffers class can be loaded at a time, and that would cause incompabilities

I have one question, can we solve the problem fundamentally with a wrapper? Even with a wrapper, arrow still has flatbuffer dependency which may incompatible with user application, right?

**Comment by wesm** [comment]
Since we're very close to the 0.14.0 release, can we reach some conclusion on this in the next couple of days?

**Comment by emkornfield** [comment]
> I have one question, can we solve the problem fundamentally with a wrapper? Even with a wrapper, arrow still has flatbuffer dependency which may incompatible with user application, right?

Not if our dependency is relocated and isn't  exposed to the public.  Then two versions can coexist but ours need to be fixed.  @tianchen92 would you like to try my suggested fix?  Otherwise I can try it out

**Comment by tianchen92** [comment]
@emkornfield Of course, would appreciate if you could give some more information about how to try this approach exactly since I am not quite sure with some details. I will have a try then, thanks!

**Comment by emkornfield** [comment]
Let me see if I can prototype something to share

**Comment by tianchen92** [comment]
> Let me see if I can prototype something to share

Thank you @emkornfield. 
Meanwhile I have some new thoughts, since I think moving related classes to format may cause some problems because seems some dependencies changed(e.g. arrow-memory) as well as generated-class in vector. I try a new way as follow:
1、Add a wrapper in format module and which used in exposed public interface
2、Relocate flatbuffers in vector module to avoid incompatible with user application.
This way seems works fine with both maven and Intellij, do you think it is reasonable? code branch is: https://github.com/tianchen92/arrow/commits/ARROW-5579-testWrapper
If this not the correct way, then hope to see your share, thanks!

**Comment by emkornfield** [comment]
> code branch is: https://github.com/tianchen92/arrow/commits/ARROW-5579-testWrapper
If this not the correct way, then hope to see your share, thanks!

I'm not convinced this branch properly shades flatbuffers.  How did you validate?

**Comment by emkornfield** [comment]
@tianchen92 I'm going to close this in favor of https://github.com/apache/arrow/pull/4701

**Pull request closed without merging**